hibernate-core/src/main/java/org/hibernate/loader/ast/internal/AbstractEntityBatchLoader.java

```diff
@@ -92,11 +96,32 @@ private T load(
 		if ( hasSingleId ) {
 			return singleIdLoader.load( id, entityInstance, lockOptions, readOnly, session );
 		}
-
-		initializeEntities( ids, id, entityInstance, lockOptions, readOnly, session );
-
-		final EntityKey entityKey = session.generateEntityKey( id, getLoadable().getEntityPersister() );
-		//noinspection unchecked
-		return (T) session.getPersistenceContext().getEntity( entityKey );
+		if ( lockOptions.getLockMode().greaterThan( LockMode.OPTIMISTIC_FORCE_INCREMENT ) ) {
+			// we want to apply the lock only to the first entity, we need to generate 2 different queries
+			ids = trimIds( 1, ids  );
+			initializeEntities( ids, ids[0], entityInstance, new LockOptions( LockMode.NONE ), readOnly, session );
+			return singleIdLoader.load( id, entityInstance, lockOptions, readOnly, session );
+		}
+		else {
+			// if LockMode is different from NONE but less than OPTIMISTIC_FORCE_INCREMENT we still want to apply it only to the first entity but we can generate 1 query and then
+			// do an upgrade lock
+			ids = trimIds( 0, ids );
+			initializeEntities( ids, id, entityInstance, new LockOptions( LockMode.NONE ), readOnly, session );
+
+			final EntityKey entityKey = session.generateEntityKey( id, getLoadable().getEntityPersister() );
+			final Object entity =  session.getPersistenceContext().getEntity( entityKey );
+			if ( lockOptions.getLockMode() != LockMode.NONE ) {
+				LoaderHelper.upgradeLock(
+						entity,
+						session.getPersistenceContext().getEntry( entity ),
+						lockOptions,
+						(EventSource) session
+				);
+			}
+
+			return (T) entity;
+		}
 	}
+
+	protected abstract Object[] trimIds(int startPosition, Object[] keysToInitialize);
```

Comment:
I guess this change isn't needed anymore if we load the full batch first and do a lock request then.